fl4health/clients/apfl_client.py

```diff
@@ -1,255 +1,149 @@
 
 import torch
-from flwr.common.logger import log
-from flwr.common.typing import Config, NDArrays, Scalar
-from torch.nn.modules.loss import _Loss
+from flwr.common.typing import Config
+from torch.optim import Optimizer
 from torch.utils.data import DataLoader
 
-from fl4health.clients.numpy_fl_client import NumpyFlClient
+from fl4health.checkpointing.checkpointer import TorchCheckpointer
+from fl4health.clients.basic_client import BasicClient
 from fl4health.model_bases.apfl_base import ApflModule
-from fl4health.utils.metrics import Metric, MetricAverageMeter, MetricMeter
+from fl4health.parameter_exchange.layer_exchanger import FixedLayerExchanger
+from fl4health.reporting.fl_wanb import ClientWandBReporter
+from fl4health.utils.losses import Losses, LossMeter, LossMeterType
+from fl4health.utils.metrics import Metric, MetricMeter, MetricMeterManager, MetricMeterType
 
-LocalLoss = torch.Tensor
-GlobalLoss = torch.Tensor
-PersonalLoss = torch.Tensor
 
-LocalPreds = torch.Tensor
-GlobalPreds = torch.Tensor
-PersonalPreds = torch.Tensor
-
-ApflTrainStepOutputs = Tuple[LocalLoss, GlobalLoss, PersonalLoss, LocalPreds, GlobalPreds, PersonalPreds]
-
-
-class ApflClient(NumpyFlClient):
+class ApflClient(BasicClient):
     def __init__(
         self,
         data_path: Path,
         metrics: Sequence[Metric],
         device: torch.device,
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        metric_meter_type: MetricMeterType = MetricMeterType.AVERAGE,
+        checkpointer: Optional[TorchCheckpointer] = None,
     ) -> None:
-        super().__init__(data_path, device)
+        super(BasicClient, self).__init__(data_path, device)
         self.metrics = metrics
-        self.model: ApflModule
+        self.checkpointer = checkpointer
+        self.train_loss_meter = LossMeter.get_meter_by_type(loss_meter_type)
+        self.val_loss_meter = LossMeter.get_meter_by_type(loss_meter_type)
+
+        # Define mapping from prediction key to meter to pass to MetricMeterManager constructor for train and val
+        train_key_to_meter_map = {
+            "local": MetricMeter.get_meter_by_type(self.metrics, metric_meter_type, "train_meter_local"),
+            "global": MetricMeter.get_meter_by_type(self.metrics, metric_meter_type, "train_meter_global"),
+            "personal": MetricMeter.get_meter_by_type(self.metrics, metric_meter_type, "train_meter_personal"),
+        }
+        self.train_metric_meter_mngr = MetricMeterManager(train_key_to_meter_map)
+        val_key_to_meter_map = {
+            "local": MetricMeter.get_meter_by_type(self.metrics, metric_meter_type, "val_meter_local"),
+            "global": MetricMeter.get_meter_by_type(self.metrics, metric_meter_type, "val_meter_global"),
+            "personal": MetricMeter.get_meter_by_type(self.metrics, metric_meter_type, "val_meter_personal"),
+        }
+        self.val_metric_meter_mngr = MetricMeterManager(val_key_to_meter_map)
```

Comment:
We don't have to do it in this PR, since we're going to follow this up with more changes to fold in the `MetricMeterManager`, but a clean way to do these constructions would be useful 🙂

Reply:
Yeah i will reserve that for the next PR since we are having the user pass MetricMeterManager and they way we construct them is likely to change